### PR TITLE
[SPARK-45480][SQL][UI] Selectable Spark Plan Node on UI

### PR DESCRIPTION
**Pull request by yaooqinn**
### What changes were proposed in this pull request?

This PR introduces selectable animation for Spark SQL Plan Node On UI, which lights up the selected node and its linked nodes and edges.

### Why are the changes needed?

Better UX for SQL plan visualization and debugging. Especially for large queries, users can now concentrate on the current node and its nearest neighbors to get a better understanding of node lineage.

### Does this PR introduce _any_ user-facing change?

Yes, let's see the video.

### How was this patch tested?

https://github.com/apache/spark/assets/8326978/f5ba884c-acce-46b8-8568-3ead55c91d4f

### Was this patch authored or co-authored using generative AI tooling?

no

### CODE REVIEW

**Comment by yaooqinn** [comment]
cc @sarutak @dongjoon-hyun @cloud-fan @HyukjinKwon @zhengruifeng @LuciferYang thanks

**Comment by zhengruifeng** [comment]
also cc @jasonli-db and @rednaxelafx

File: sql/core/src/main/resources/org/apache/spark/sql/execution/ui/static/spark-sql-viz.js [hunk 1]
```diff
@@ -269,3 +270,47 @@ function togglePlanViz() { // eslint-disable-line no-unused-vars
...
+        .classed("selected", false)
+        .classed("linked", false);
+      d3.select("#" + selectNode + " rect").classed("selected", true);
+      linkedNodes.forEach((linkedNode) => {
+        d3.select("#" + linkedNode + " rect").classed("linked", true);
+      });
+      linkedEdges.get(selectNode).forEach((linkedEdge) => {
+        const arrowHead = d3.select("#" + linkedEdge + " path");
+        arrowHead.classed("linked", true);
+        const arrowShaft = $(arrowHead.node()).parents("g.edgePath").children("path");
```
Comment (LuciferYang) [inline sql/core/src/main/resources/org/apache/spark/sql/execution/ui/static/spark-sql-viz.js:304]: How about using d3's api uniformly? 
Like `d3.select(arrowHead.node().parentNode).select("path");` ?

File: sql/core/src/main/resources/org/apache/spark/sql/execution/ui/static/spark-sql-viz.js [hunk 2]
```diff
@@ -269,3 +270,47 @@ function togglePlanViz() { // eslint-disable-line no-unused-vars
...
+    const edge = g.edge(e);
+    const from = g.node(e.v);
+    const to = g.node(e.w);
+    collectLinks(linkedNodes, from.id, to.id);
+    collectLinks(linkedNodes, to.id, from.id);
+    collectLinks(linkedEdges, from.id, edge.arrowheadId);
+    collectLinks(linkedEdges, to.id, edge.arrowheadId);
+  });
+
+  linkedNodes.forEach((linkedNodes, selectNode) => {
```
Comment (LuciferYang) [inline sql/core/src/main/resources/org/apache/spark/sql/execution/ui/static/spark-sql-viz.js:291]: It appears that each click currently clears and resets the class of all nodes and edges. Is it possible to update only those nodes and edges that actually change? Please correct me if I'm wrong.

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM. Thank you for the nice improvement, @yaooqinn .

File: sql/core/src/main/resources/org/apache/spark/sql/execution/ui/static/spark-sql-viz.js (hunk 2 above)
Comment (LuciferYang) [reply sql/core/src/main/resources/org/apache/spark/sql/execution/ui/static/spark-sql-viz.js:291]: Ok, let's consider how to optimize when there is a real bottleneck.

File: sql/core/src/main/resources/org/apache/spark/sql/execution/ui/static/spark-sql-viz.js (hunk 1 above)
Comment (yaooqinn) [reply sql/core/src/main/resources/org/apache/spark/sql/execution/ui/static/spark-sql-viz.js:304]: The head and shaft are not at the same level, in order to get their common ancestor with the d3 selection API, we need to hard code like
```javascript
d3.select(d3.select(arrowHead.node().parentNode).node().parentNode).select("path")
```

It appears not future-proofing. The suitable way is using edge ID, but it seems not supported by both graphdot-lib and dagre-d3 yet. In this scenario, jQuery appears to be a suitable option based on our typical usage in UI.

**Comment by rednaxelafx** [comment]
A nice improvement to the UI 👍 

But there's a dilemma: this feature is supposed to improve the UI's usability on large plans. For us it's quite common to have very large query plan graphs and the existing UI is already quite heavy (takes a while to render...).
In that case this new feature might actually make the usability worse when a user clicks on a node and triggers a whole-graph update.

Can we have either a Spark conf or a checkbox on the UI to allow users to opt-out of this feature when they find it counterproductive in their scenario?

Side note: when I used to work on the HotSpot C2 compiler, I used the IdealGraphVisualizer (IGV) a lot to visualize the C2 IR. It also provides node selection and highlighting capabilities, and in addition it supports "node hiding" -- you can choose to focus on a subset of nodes and hide all of the other nodes, and optionally show the immediate neighbor nodes of the selected subgraph to be shown in lower opacity, and double-clicking on these neighbors will add them to the selection.

It was super useful and my workflow was heavily dependent on that. I'd start out from the specific node in question and gradually expand out to cover enough context, and try to figure out what the problem was. In a graph of thousands of nodes this allowed me to work with mostly a couple dozen nodes at a time.

**Comment by sarutak** [comment]
@yaooqinn Looks like a nice feature!

> Can we have either a Spark conf or a checkbox on the UI to allow users to opt-out of this feature when they find it counterproductive in their scenario?

+1

**Comment by yaooqinn** [comment]
> In that case this new feature might actually make the usability worse when a user clicks on a node and triggers a whole-graph update.

Hi @rednaxelafx @sarutak  and @LuciferYang, thanks for the check. I have avoided the whole plan updates in [31cff1f](https://github.com/apache/spark/pull/43307/commits/31cff1fced1f38423c240993891b27c128e77121). Note that we do not rerender the whole plan after click, is there any reason to add a checkbox here.

Thanks for the side note @rednaxelafx, I think I'll go and experience it and bring some interesting things back to Spark UI.

**Review by sarutak — APPROVED** [review]

**Review by LuciferYang — APPROVED** [review]

**Review by HyukjinKwon — APPROVED** [review]

**Comment by yaooqinn** [comment]
Thanks @dongjoon-hyun @sarutak @LuciferYang @HyukjinKwon and @rednaxelafx 

Merged to master

**Pull request closed without merging**